Improvements to the S3 sink documenation (#6383)
opensearch-project/documentation-website: _data-prepper/pipelines/configuration/sinks/s3.md

# s3

> The `s3` sink saves batches of events to [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) objects and writes batches of Data Prepper events to S3 objects. The configured `codec` determines how the `s3` sink serializes the data into S3.

Reviewer: ```suggestion
The `s3` sink saves and writes batches of Data Prepper events to Amazon Simple Storage Service (Amazon S3) objects. The configured `codec` determines how the `s3` sink serializes the data into Amazon S3.
```

> The `s3` sink saves batches of events to [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) objects and writes batches of Data Prepper events to S3 objects. The configured `codec` determines how the `s3` sink serializes the data into S3.

Author: @natebower , I think those changes make sense. And yes, these are "Data Prepper events."

The `s3` sink uses the following format when batching events:

```
${pathPrefix}events-%{yyyy-MM-dd'T'HH-mm-ss'Z'}-${currentTimeInNanos}-${uniquenessId}.${codecSuppliedExtension}
```

When a batch of objects is written to S3, the format of the object looks similar to the following:

```
my-logs/2023/06/09/06/events-2023-06-09T06-00-01-1686290401871214927-ae15b8fa-512a-59c2-b917-295a0eff97c8.json
```

> For more information on how to configure an object, see the [Object key](#object-key-configuration) section.

Reviewer: ```suggestion
For more information about how to configure an object, see the [Object key](#object-key-configuration) section.
```

        aws:
          region: us-east-1
          sts_role_arn: arn:aws:iam::123456789012:role/Data-Prepper
        max_retries: 5
        bucket: mys3bucket
        object_key:
          path_prefix: my-logs/%{yyyy}/%{MM}/%{dd}/
        threshold:
          event_count: 10000
          maximum_size: 50mb
          event_collect_timeout: 15s
        codec:
          ndjson:
        buffer_type: in_memory
```

## IAM permissions

> In order to use the `s3` sink, configure your AWS Identity and Access Management (IAM) permissions to grant Data Prepper access to write to Amazon S3. You can use a configuration similar to the following JSON configuration:

Reviewer: This works for me.

> In order to use the `s3` sink, configure your AWS Identity and Access Management (IAM) permissions to grant Data Prepper access to write to Amazon S3. You can use a configuration similar to the following JSON configuration:

Reviewer: ```suggestion
In order to use the `s3` sink, configure AWS Identity and Access Management (IAM) to grant Data Prepper permissions to write to Amazon S3. You can use a configuration similar to the following JSON configuration:
```

`sts_role_arn` | No | String | The AWS Security Token Service (AWS STS) role to assume for requests to Amazon SQS and Amazon S3. Defaults to `null`, which will use the [standard SDK behavior for credentials](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html).
`sts_header_overrides` | No | Map | A map of header overrides that the IAM role assumes for the sink plugin.
`sts_external_id` | No | String | An STS external ID to use when Data Prepper assumes the role. For more information, see the documentation for `ExternalId` in the [STS AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) reference.

## Threshold configuration

Use the following options to set ingestion thresholds for the `s3` sink. When any of these conditions is met, Data Prepper will write events to an S3 object.

Option | Required | Type | Description
:--- | :--- | :--- | :---
`event_count` | Yes | Integer | The number of Data Prepper events to accumulate before writing an object to S3.

> `maximum_size` | No | String | The maximum number of bytes to accumulate before writing an object to S3. Defaults to `50mb`.

Reviewer 2: ```suggestion
`maximum_size` | No | String | The maximum number of bytes to accumulate before writing an object to S3. Default is `50mb`.
```

> `event_collect_timeout` | Yes | String | The maximum timeout before Data Prepper writes an event to S3. The value should be either an ISO-8601 duration, such as `PT2M30S`, or a simple notation, such as `60s` or `1500ms`.

Reviewer: ```suggestion
`event_collect_timeout` | Yes | String | The maximum amount of time before Data Prepper writes an event to S3. The value should be either an ISO-8601 duration, such as `PT2M30S`, or a simple notation, such as `60s` or `1500ms`.
```

> `event_collect_timeout` | Yes | String | The maximum timeout before Data Prepper writes an event to S3. The value should be either an ISO-8601 duration, such as `PT2M30S`, or a simple notation, such as `60s` or `1500ms`.

Reviewer 2: Something like "The maximum amount of time allowed before"?

Reviewer: If we wanted to still use the word, "timeout", could we do, "The maximum duration before Data Prepper timesout when writing an event to S3"?

Reviewer: That would change the technical meaning wouldn't it @dlvenable? Is "timeout duration" the length of time that Data Prepper has to write an event to S3 before timing out, or is the "timeout duration" the length of the time we set for Data Prepper to prepare to write an event? If its the second one, I feel like we should go @natebower original suggestion, since that technically doesn't fit the definition of "timeout" despite the word being used in the setting.

Author: @natebower , The term "timeout" is not really ideal either in the configuration or in the description. Nothing is really "timing out" here. This is really more of a flush interval. 

> or is the "timeout duration" the length of the time we set for Data Prepper to prepare to write an event? If its the second one, I feel like we should go @natebower original suggestion, since that technically doesn't fit the definition of "timeout" despite the word being used in the setting.

Yes, it is the second one. The amount of time before Data Prepper writes the event.

Perhaps we can say "The maximum amount of time before".

Reviewer: Let's go with the "The maximum amount of time before."

## Buffer type

> `buffer_type` is an optional configuration that determines how Data Prepper temporarily holds data before writing an S3 object to S3. The default value is `in_memory`. Use one of the following options:

Reviewer 2: ```suggestion
`buffer_type` is an optional configuration that determines how Data Prepper temporarily stores data before writing an object to S3. The default value is `in_memory`. Use one of the following options:
```

- `in_memory`: Stores the record in memory.

> - `local_file`: Flushes the record into a file on your machine. This uses your local machine's temporary directory.

Reviewer 2: ```suggestion
- `local_file`: Flushes the record into a file on your local machine. This uses your machine's temporary directory.
```

- `multipart`: Writes using the [S3 multipart upload](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html). Every 10 MB is written as a part.

## Object key configuration

Option | Required | Type | Description
:--- | :--- | :--- | :---

> `path_prefix` | No | String | The S3 key prefix path to use for objects written to S3. Accepts date-time formatting. For example, you can use `%{yyyy}/%{MM}/%{dd}/%{HH}/` to create hourly folders in S3. You should have this end with a `/`. By default, Data Prepper writes objects to the root of the S3 bucket.

Reviewer: ```suggestion
`path_prefix` | No | String | The S3 key prefix path to use for objects written to S3. Accepts date-time formatting. For example, you can use `%{yyyy}/%{MM}/%{dd}/%{HH}/` to create hourly folders in S3. The prefix path should end with `/`. By default, Data Prepper writes objects to the root of the S3 bucket.
```

> `path_prefix` | No | String | The S3 key prefix path to use for objects written to S3. Accepts date-time formatting. For example, you can use `%{yyyy}/%{MM}/%{dd}/%{HH}/` to create hourly folders in S3. You should have this end with a `/`. By default, Data Prepper writes objects to the root of the S3 bucket.

Reviewer 2: Third sentence: Let's revise to "The [noun] should end with a `/`."

Author: "The path prefix should end with a `/`."

Author: @natebower , See the comment above.

`schema` | Yes | String | The Avro [schema declaration](https://avro.apache.org/docs/current/specification/#schema-declaration). Not required if `auto_schema` is set to true.
`auto_schema` | No | Boolean | When set to `true`, automatically generates the Avro [schema declaration](https://avro.apache.org/docs/current/specification/#schema-declaration) from the first event.

### Setting a schema with Parquet

> The following example shows you how to configure the `s3` sink to write Parquet data into a Parquet file using a schema for VPC Flow Logs:

Reviewer: ```suggestion
The following example shows you how to configure the `s3` sink to write Parquet data into a Parquet file using a schema for [VPC Flow Logs](https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs.html#flow-log-records):
```

`schema` | Yes | String | The Avro [schema declaration](https://avro.apache.org/docs/current/specification/#schema-declaration). Not required if `auto_schema` is set to true.
`auto_schema` | No | Boolean | When set to `true`, automatically generates the Avro [schema declaration](https://avro.apache.org/docs/current/specification/#schema-declaration) from the first event.

### Setting a schema with Parquet

> The following example shows you how to configure the `s3` sink to write Parquet data with a schema for VPC Flow Logs:

Reviewer: ```suggestion
The following example shows you how to configure the `s3` sink to write Parquet data into a Parquet file using a schema for VPC Flow Logs:
```

Reviewer: @natebower: Does my suggestion help?

> The following example shows how you can configure the `s3` sink to write Parquet data with a schema for the AWS VPC Flow logs.

Reviewer 2: ```suggestion
The following example shows you how to configure the `s3` sink to write Parquet data with a schema for VPC Flow Logs:
```

> The following example shows how you can configure the `s3` sink to write Parquet data with a schema for the AWS VPC Flow logs.

Reviewer 2: The meaning after "sink" is a little unclear. Should we say to where the data is being written? Also clarify what we mean by "with a schema for". I assumed that we are referring to VPC Flow Logs, the Amazon VPC feature.

Author: @natebower , How about this?

> The `s3` sink can write data in Parquet format using a schema that you specify. The following example shows how you can do this. The example creates a schema based on the AWS VPC Flow logs data model.

Yes, this is the Amazon VPC feature. It is documented here: https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs.html#flow-log-records
